src/vcpkg/base/downloads.cpp

```diff
@@ -672,107 +132,168 @@ namespace vcpkg
+            curl_easy_setopt(curl, CURLOPT_PRIVATE, &data);
+            if (outputs.size() > request_index)
             {
-                if (!new_curl_seen)
+                const auto& output = outputs[request_index];
+
+                std::error_code ec;
+                data.file.reset(new WriteFilePointer(output, Append::NO, ec));
+                if (ec)
                 {
-                    // old version of curl, we only have the result code for the last operation
-                    context.report_error(msgCurlFailedGeneric, msg::exit_code = *this_batch_exit_code);
+                    context.report_error(format_filesystem_call_error(ec, "fopen", {output}));
                 }
-
-                if (ret.size() != last_try_op)
+                else
                 {
-                    // curl didn't process everything we asked of it; this usually means curl crashed
-                    auto command_line = std::move(batch_cmd).extract();
-                    replace_secrets(command_line, secrets);
-                    context.report_error_with_log(Strings::join("\n", debug_lines),
-                                                  msgCurlFailedToReturnExpectedNumberOfExitCodes,
-                                                  msg::exit_code = *this_batch_exit_code,
-                                                  msg::command_line = command_line);
-                    return ret;
+                    curl_easy_setopt(curl, CURLOPT_WRITEDATA, data.file.get());
+                    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &write_file_callback);
                 }
             }
-            else
+            curl_multi_add_handle(multi_handle, curl);
+        }
+
+        int still_running = 0;
+        do
+        {
+            CURLMcode mc = curl_multi_perform(multi_handle, &still_running);
+            if (mc != CURLM_OK)
+            {
+                context.report_error(msg::format(msgCurlFailedGeneric,
+                                                 msg::exit_code = static_cast<int>(mc),
+                                                 msg::error_msg = curl_multi_strerror(mc)));
+            }
+
+            mc = curl_multi_poll(multi_handle, nullptr, 0, 1000, nullptr);
+            if (mc != CURLM_OK)
             {
-                // couldn't even launch curl, record this as the last fatal error and give up
-                return ret;
+                context.report_error(msg::format(msgCurlFailedGeneric,
+                                                 msg::exit_code = static_cast<int>(mc),
+                                                 msg::error_msg = curl_multi_strerror(mc)));
+            }
+        } while (still_running);
+
+        int messages_in_queue = 0;
+        while (auto* msg = curl_multi_info_read(multi_handle, &messages_in_queue))
+        {
+            // just drain any messages left in the queue from the previous loop
```

Comment:
This comment does not seem to match what the loop above does as it is extracting exit codes and similar while the loop above does not do that at all. Should the loop above be doing that?